WIP: Parametric ReLU

creatorscan opened the pull request:
Hi Karel,
Sorry for taking too long and  thanks for your patience. 
I have updated the files based on the new Kaldi version. Please review the changes and let me know.

Commit:
modified:   src/cudamatrix/cu-kernels-ansi.h
	modified:   src/cudamatrix/cu-kernels.cu
	modified:   src/cudamatrix/cu-kernels.h
	modified:   src/cudamatrix/cu-matrix.cc
	modified:   src/cudamatrix/cu-matrix.h
	modified:   src/matrix/kaldi-matrix.cc
	modified:   src/matrix/kaldi-matrix.h
	modified:   src/nnet/nnet-component.cc
	modified:   src/nnet/nnet-component.h
	src/nnet/nnet-parametric-relu.h

KarelVesely84 commented on src/cudamatrix/cu-kernels.cu:
```diff
@@ -1763,6 +1763,31 @@ static void _group_max(Real *y, const Real *x, MatrixDim d, int src_stride,
 
+template<typename Real>
+__global__
+static void _parametric_relu(Real*y, const Real*x, MatrixDim d, int src_stride, const Real*a, const Real*b) {
```
Hi, I'll be highlighting the code, pointing out at some details od the coding style we use.
The basic rule is that the style of the code should be the same as elsewhere in the file, 
including the white-spaces: 
'Real_y' -> 'Real *y'
'j_d.stride' -> 'j \* d.stride'

(and also there should be no whitespace at the end of the lines)

KarelVesely84 commented on src/cudamatrix/cu-kernels.cu:
```diff
@@ -1763,6 +1763,31 @@ static void _group_max(Real *y, const Real *x, MatrixDim d, int src_stride,
+static void _parametric_relu(Real*y, const Real*x, MatrixDim d, int src_stride, const Real*a, const Real*b) {
+  int i = blockIdx.x * blockDim.x + threadIdx.x;
+  int j = blockIdx.y * blockDim.y + threadIdx.y;
+  int dst_index = i + j*d.stride, src_index = i + j * src_stride;
```
It's better to put `src_index` to next line under the `dst_index`, this would be more readable.

KarelVesely84 commented on src/cudamatrix/cu-kernels.cu:
```diff
@@ -1763,6 +1763,31 @@ static void _group_max(Real *y, const Real *x, MatrixDim d, int src_stride,
+  int i = blockIdx.x * blockDim.x + threadIdx.x;
+  int j = blockIdx.y * blockDim.y + threadIdx.y;
+  int dst_index = i + j*d.stride, src_index = i + j * src_stride;
+  if(i < d.cols && j < d.rows) {
```
There should be space between 'if' and '(', it is part of the Google style. The script which checks the style is here: https://github.com/kaldi-asr/kaldi/blob/master/src/doc/cpplint.py

KarelVesely84 commented on src/cudamatrix/cu-kernels.cu:
```diff
@@ -1763,6 +1763,31 @@ static void _group_max(Real *y, const Real *x, MatrixDim d, int src_stride,
+  int e_index   = i + j*e_stride;
+  int y_index   = i + j*y_stride;
+  if (i < d.cols  && j < d.rows )
+    eout[dst_index] = (y[y_index] > 0.0 ? a[i] * e[e_index] : b[i] * e[e_index]);
```
It would be better to do the decision according to the value of 'x' rather than 'y'. In the forward pass the decision was also based on 'x'.

creatorscan replied:
I used 'y' to denote we check activation 'out'->'y' is checked for < or > 0. I guess, it will be better to use 'y' than 'x'.

KarelVesely84 replied:
Aha, I see. I would also expect that `x` is corresponding to `in` and `y` to `out`. I was thinking there could be a possible situation in which `alpha` or `beta` becomes negative. Then it would be crucial to use `in` -> `x`.

But still ... , If the `PropagateFnc` is doing the decision based on `x` it feels natural that the decision in `BackpropagateFnc` is also based on `x`.

KarelVesely84 commented on src/cudamatrix/cu-matrix.cc:
```diff
@@ -1229,6 +1229,47 @@ void CuMatrixBase<Real>::AddMatMatElements(Real alpha,
 
+template<typename Real>
+void CuMatrixBase<Real>::ParametricReLU(const CuMatrixBase<Real> &src, const CuVectorBase<Real> &a, const CuVectorBase<Real> &b) {
+  KALDI_ASSERT(SameDim(*this, src));
```
Please add asserts:
KALDI_ASSERT(a.Dim() == this->NumCols());
KALDI_ASSERT(b.Dim() == this->NumCols());

KarelVesely84 commented on src/cudamatrix/cu-matrix.cc:
```diff
@@ -1229,6 +1229,47 @@ void CuMatrixBase<Real>::AddMatMatElements(Real alpha,
 }
 
+template<typename Real>
+void CuMatrixBase<Real>::ParametricReLU(const CuMatrixBase<Real> &src, const CuVectorBase<Real> &a, const CuVectorBase<Real> &b) {
```
Please rename the method to 'ParametricRelu' (it's consistent with other methods).

KarelVesely84 commented on src/cudamatrix/cu-matrix.cc:
```diff
@@ -1229,6 +1229,47 @@ void CuMatrixBase<Real>::AddMatMatElements(Real alpha,
+}
+
+template<typename Real> 
+void CuMatrixBase<Real>::DiffParametricReLU(const CuMatrixBase<Real> &value,
```
The indent of line 1255 is unusual. With such long function names it is fine to put one argument per line on top of each other (ideally the lines should not be longer than 80 characters).

KarelVesely84 commented on src/matrix/kaldi-matrix.cc:
```diff
@@ -2717,6 +2717,45 @@ void MatrixBase<Real>::AddToRows(Real alpha, Real *const *dst) const {
+  const Real *src_data = src.data_;
+  const Real *adata = a.Data();
+  const Real *bdata = b.Data();
+   for (MatrixIndexT r = 0; r < num_rows; r++) {
```
Line 2729 should have no indent...

KarelVesely84 commented on src/matrix/kaldi-matrix.cc:
```diff
@@ -2717,6 +2717,45 @@ void MatrixBase<Real>::AddToRows(Real alpha, Real *const *dst) const {
+
+template<typename Real>
+void MatrixBase<Real>::DiffParametricReLU(const MatrixBase<Real> &value,
+                                   const MatrixBase<Real> &diff, const VectorBase<Real> &a, const VectorBase<Real> &b) {
```
Please, put one argument per line...

KarelVesely84 commented on src/matrix/kaldi-matrix.cc:
```diff
@@ -2717,6 +2717,45 @@ void MatrixBase<Real>::AddToRows(Real alpha, Real *const *dst) const {
+
+
+template<typename Real>
+void MatrixBase<Real>::DiffParametricReLU(const MatrixBase<Real> &value,
```
The name argument 'value' does not say which matrix should be used. It is true that it is consistent with 'DiffSigmoid' and 'DiffTanh', and I saw the comment at the declaration... Okay, let's keep it as it is...

vesis84 commented:
Hi, I fininshed the review. I cannot see the file 'nnet/nnet-parametric-relu.h' although it says the file was committed, this is strange. Also this pull request has contain many commits from other people. We could try to 'rebase' and do a force push to your branch. This is not the cleanest solution, but it should resolve the problem, which arised from creating a pull-request against the repo which was outdated.

Commit:
modified:   src/cudamatrix/cu-kernels.cu
	modified:   src/cudamatrix/cu-matrix.cc
	modified:   src/cudamatrix/cu-matrix.h
	modified:   src/matrix/kaldi-matrix.cc
	modified:   src/matrix/kaldi-matrix.h
	modified:   src/nnet/nnet-component.cc
	modified:   src/nnet/nnet-component.h
	modified:   src/nnet/nnet-parametric-relu.h

Commit:
new file:   src/nnet/nnet-parametric-relu.h

Commit:
new file:   src/nnet/nnet-parametric-relu.h

creatorscan commented:
Karel,
It was my mistake as nnet-parametric-relu.h was an Untracked file. I just added it now in my latest commit.

KarelVesely84 commented on src/nnet/nnet-parametric-relu.h:
```diff
@@ -0,0 +1,284 @@
+// nnet/nnet-parametric-relu.h
+
+// Copyright 2011-2014  Brno University of Technology (author: Karel Vesely)
```
Please put 2016 and your name as author...

KarelVesely84 commented on src/nnet/nnet-parametric-relu.h:
```diff
@@ -0,0 +1,284 @@
+ public:
+  ParametricRelu(int32 dim_in, int32 dim_out):
+    UpdatableComponent(dim_in, dim_out),
+    max_alpha_(dim_out), min_beta_(dim_out),
```
One initialized variable per line is 'nicer' and more readable...

KarelVesely84 commented on src/nnet/nnet-parametric-relu.h:
```diff
@@ -0,0 +1,284 @@
+  }
+
+ private:
+  CuVector<BaseFloat> max_alpha_;
```
I would be good to write the purpose of the variables in a comment... It can look like this:
BaseFloat max_alpha_;  ///< Vector of 'alphas', one value per neuron.

KarelVesely84 replied:
Then, there are quite many variables, some of the may be redundant?

KarelVesely84 commented on src/nnet/nnet-parametric-relu.h:
```diff
@@ -0,0 +1,284 @@
+      else if (token == "<MaxLearnRateCoef>")  ReadBasicType(is, false, &max_learn_rate_coef);
+      else if (token == "<MinLearnRateCoef>")  ReadBasicType(is, false, &min_learn_rate_coef);
+      else if (token == "<FixMaxCoef>")  ReadBasicType(is, false, &max_fix_coef);
+      else if (token == "<FixMinCoef>")   ReadBasicType(is, false, &min_fix_coef);
```
FixMinCoef, there is 2x the same line...

Also it would be good if the `<Tokens>` and `variables` had the same name. The name should be well chosen to be minimalistic and to characterize its purpose.

KarelVesely84 commented on src/nnet/nnet-parametric-relu.h:
```diff
@@ -0,0 +1,284 @@
+       min_beta_corr_.AddRowSumMat(1.0, input_min_, mmt);
+       min_beta_.AddVec(-blr, min_beta_corr_);
+    }
+    if (clip_gradient_ > 0.0) {  // gradient clipping
```
It would be better to apply the clipping immediately before the update of the parameters...

KarelVesely84 commented on src/nnet/nnet-parametric-relu.h:
```diff
@@ -0,0 +1,284 @@
+  void Update(const CuMatrixBase<BaseFloat> &input,
+              const CuMatrixBase<BaseFloat> &diff) {
+    // we use following hyperparameters from the option class
+    const BaseFloat alr = opts_.learn_rate * max_learn_rate_coef_;
```
It would be better to replace `max_learn_rate_coef_` and `min_learn_rate_coef_` by the variables `bias_learn_rate_coef_` (alphas) and `learn_rate_coef_` (betas) from the interface `UpdatableComponent::`. The benefit is that these can be adjusted externally with the tool `nnet-set-learnrate`.

KarelVesely84 commented on src/nnet/nnet-parametric-relu.h:
```diff
@@ -0,0 +1,284 @@
+  }
+
+  void PropagateFnc(const CuMatrixBase<BaseFloat> &in,
+                    CuMatrixBase<BaseFloat> *out) {
```
It would be good to add a comment with the formula and short description.

```
// Multiply activations by ReLU scalars (alpha, beta):
// out = in * (in >= 0.0 ? alpha : beta)
```

KarelVesely84 commented on src/nnet/nnet-parametric-relu.h:
```diff
@@ -0,0 +1,284 @@
+                        const CuMatrixBase<BaseFloat> &out,
+                        const CuMatrixBase<BaseFloat> &out_diff,
+                        CuMatrixBase<BaseFloat> *in_diff) {
+    // multiply error derivative by activations (alpha, beta)
```
```
// Multiply error derivative by ReLU scalars (alpha, beta):
// in_diff = out_diff * (in >= 0.0 ? alpha : beta)
```

KarelVesely84 commented on src/nnet/nnet-parametric-relu.h:
```diff
@@ -0,0 +1,284 @@
+    const BaseFloat mmt = opts_.momentum;
+    // compute gradient (incl. momentum)
+    if (!max_fix_coef_) {  // the alpha parameter is learnable
+       input_max_.Resize(input.NumRows(), input.NumCols());
```
This could be shortened as:

```
input_max_ = input;
```

KarelVesely84 commented on src/nnet/nnet-parametric-relu.h:
```diff
@@ -0,0 +1,284 @@
+       input_max_.Resize(input.NumRows(), input.NumCols());
+       input_max_.CopyFromMat(input);
+       input_max_.ApplyFloor(0.0);
+       max_alpha_corr_.AddRowSumMat(1.0, input_max_, mmt);
```
The derivative is missing the term `dE / d_out` from the chain rule. To make it correct, we should add there this line:

```
input_max_.MulElements(diff);
```

KarelVesely84 commented on src/nnet/nnet-parametric-relu.h:
```diff
@@ -0,0 +1,284 @@
+       input_min_.Resize(input.NumRows(), input.NumCols());
+       input_min_.CopyFromMat(input);
+       input_min_.ApplyCeiling(0.0);
+       min_beta_corr_.AddRowSumMat(1.0, input_min_, mmt);
```
Symmetrically, it applies here as well:

```
input_min_.MulElements(diff);
```

vesis84 commented:
Hi, thank you for the update. I'll take a break for a moment. This module will probably take some time to refactor... It is good that we found the bug in the `Update()` function, let's see the updated results :)

creatorscan commented:
Thanks a lot Karel for pointing it out.
When I updated the code I forgot to add the 'MulElements'. Anyways, I will
incorporate your reviewed changes and let you know once you are free.

Thank you

On Wed, 25 May 2016 at 14:42 Karel Vesely <notifications@github.com> wrote:

> Hi, thank you for the update. I'll take a break for a moment. This module
> will probably take some time to refactor... It is good that we found the
> bug in the Update() function, let's see the updated results :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/vesis84/kaldi/pull/3#issuecomment-221563320>
>

Commit:
modified:   src/nnet/nnet-parametric-relu.h

vesis84 commented:
Hi, sorry for the delay, I just noticed the code is updated. 
Could you please rebase with `vesis84:master`? 

It should be something like:
```
git remote add vesis84 ...URL...
git fetch vesis84 vesis84/master
git rebase vesis84/master
git push origin Parametric-ReLU
```

Commit:
modified:   src/cudamatrix/cu-kernels-ansi.h
	modified:   src/cudamatrix/cu-kernels.cu
	modified:   src/cudamatrix/cu-kernels.h
	modified:   src/cudamatrix/cu-matrix.cc
	modified:   src/cudamatrix/cu-matrix.h
	modified:   src/matrix/kaldi-matrix.cc
	modified:   src/matrix/kaldi-matrix.h
	modified:   src/nnet/nnet-activation.h
	modified:   src/nnet/nnet-component.cc
	modified:   src/nnet/nnet-component.h
	modified:   src/nnet/nnet-nnet.cc
	modified:   src/nnet/nnet-nnet.h
	modified:   src/nnet/nnet-parametric-relu.h

creatorscan commented:
Karel,
After rebasing, I face so many issues while pushing back the changes to the branch. It shows:
conflicts in two files and I am not track it down. Can you lend me some help here ?

CONFLICT (content): Merge conflict in src/cudamatrix/cu-kernels.cu

vesis84 commented:
Ohhh, this will need a bit of improvisation. Should I try to fix it?

vesis84 commented:
I think the cause of the problem was that the PR was created against my personal fork. If 'kaldi-asr/master' was used instead, the situation would be more 'standard'. Sorry I did not expect that...

creatorscan commented:
Can we commit it to Kaldi-asr itself ?

vesis84 commented:
I am still working on it. Have you been doing something with the permissions of the files? Seems like one of the commits contains change of permissions, which I need to 'un-do' somehow... https://github.com/vesis84/kaldi/pull/3/files

vesis84 commented:
Hmm, the conflict symbols like 
```
>>>>>>>>>>>
===========
<<<<<<<<<<<

```
happened to contaminate the repository. Ohhhhhhh, this is hell on earth!!!!!!
